src/ring_jwt_middleware/core.clj

```diff
@@ -74,84 +100,100 @@
+                         (hr-duration (* 1000 jwt-max-lifetime-in-sec)))
+                 err-metas)
           (pos? expired-lifetime-secs)
-          (format "This JWT max lifetime has expired %s ago"
-                  (hr-duration (* 1000 expired-lifetime-secs)))))
-      (format "This JWT doesn't contain the following fields %s"
-              (pr-str (set/difference required-fields jwt-keys))))))
+          (->err :jwt_expired
+                 (format "This JWT max lifetime has expired %s ago"
+                         (hr-duration (* 1000 expired-lifetime-secs)))
+                 err-metas)
+          :else (->pure :ok)))
+      (->err :jwt_missing_field
+             (format "This JWT doesn't contain the following fields %s"
+                     (pr-str (set/difference required-fields jwt-keys)))
+             {:jwt jwt}))))
 
 (defn default-error-handler
-  "Return an `unauthorized` HTTP response
-  and log the error along debug infos"
-  [error-msg infos]
-  (let [err {:error :invalid_jwt
-             :error_description error-msg}]
-    (log/info error-msg (pr-str (into infos err)))
-    (resp/unauthorized err)))
+  "Return an `unauthorized` HTTP response and log the error along debug infos"
+  [{:keys [error error_description] :as error-data}]
+  (log/infof "%s: %s %s" error error_description (dissoc error-data :error :error_description :raw_jwt))
+  (resp/unauthorized (dissoc error-data :raw_jwt)))
 
-(def default-jwt-lifetime-in-sec 86400)
+(def default-jwt-lifetime-in-sec
+  "Default JWT lifetime is 24h"
+  86400)
 
-(def no-revocation-strategy (constantly false))
+(def no-revocation-strategy
+  "The default function used for `:is-revoked-fn` configuration"
+  (constantly false))
 
-(defn validate-jwt
+(s/defn validate-jwt :- (result-of s/Keyword)
   "Run both expiration and user checks,
   return a vec of errors or nothing"
-  ([raw-jwt
+  ([raw-jwt :- s/Str
     jwt
-    jwt-max-lifetime-in-sec
-    jwt-check-fn
-    log-fn]
-   (let [exp-vals [(check-jwt-expiry jwt
-                                     (or jwt-max-lifetime-in-sec
-                                         default-jwt-lifetime-in-sec))]
-         checks (if (fn? jwt-check-fn)
-                  (or (try (seq (jwt-check-fn raw-jwt jwt))
-                           (catch Exception e
-                             (log-fn "jwt-check-fn thrown an exception on"
-                                     {:level :error
-                                      :jwt jwt})
-                             (throw e)))
-                      [])
-                  [])]
-     (seq (remove nil?
-                  (concat checks exp-vals)))))
-
-  ([raw-jwt jwt jwt-max-lifetime-in-sec log-fn]
-   (validate-jwt raw-jwt jwt jwt-max-lifetime-in-sec nil log-fn)))
+    jwt-max-lifetime-in-sec :- s/Num
+    jwt-check-fn]
+   (let-either [_ (check-jwt-expiry jwt (or jwt-max-lifetime-in-sec default-jwt-lifetime-in-sec))]
+     (if (fn? jwt-check-fn)
```

Comment:
Could we loosen the spec to allow vars and multimethods and such? I see there's an error log if it's not a fn so it doesn't seem urgent.